server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -107,13 +108,23 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
     // moving the temporary directory to the final destination, once success the potentially concurrent push operations
     // will be failed and will read the descriptor.json created by current push operation directly
     FileUtils.forceMkdir(outDir.getParentFile());
-    try {
-      Files.move(tmpOutDir.toPath(), outDir.toPath());
-    }
-    catch (FileAlreadyExistsException e) {
-      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.", outDir);
-      FileUtils.deleteDirectory(tmpOutDir);
-      return jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
+    if (replaceExisting) {
+      try {
+        FileUtils.cleanDirectory(outDir);
```

Comment:
This has a race where it potentially deletes data, if the process crashes in the middle of `FileUtils.cleanDirectory`. It would be bad in the case where the first kafka replica succeeds, and the second one crashes here and deletes its data. So it'd be better to overwrite the files one-by-one, by first writing them to disk as tmp files, and then moving those tmp files on top of the previously existing files. Ideally the files should be `force()`ed before being moved, since then we confirm they are fsynced to disk, and aren't left with another possible source of corruption after a crash. I would want to fsync the directory too but I don't know how to do that in Java…

Reply:
Good catch, thank you